### ISSUE TITLE
Cloning vehicles

### ISSUE THREAD

**Issue by Razip** [OpenRCT2/OpenLoco#122]
Currently, if you want an extra vehicle to follow a certain route, you'd have to buy a new vehicle, add cars if it's a train, then make orders then make it run, but if there were a cloning option, you'd just clone it and make it run.

That feature would be especially useful if your vehicle has a lot of orders or cars because it'd require a lot of clicking if you do it the normal way.

**Comment by IntelOrca** [comment]
You do not need to set the orders again, you can copy them from another vehicle.

**Comment by jgottula** [comment]
> You do not need to set the orders again, you can copy them from another vehicle. This is demonstrated in one of the tutorials.

While this is true, clicking the bottom of the other vehicle's order list **doesn't** copy the Local/Express option, which I find annoying. That might be intentional though, since the click-bottom-of-list operation merely **appends** the other vehicle's orders to the current vehicle (just as clicking an individual order **appends** that single order to the target vehicle); it doesn't **replace** the current vehicle's order list.

Perhaps this means that we should really just add a *true* "copy order list from other vehicle" button to the vehicle dialog route tab UI, since:

- It would be more intuitive than the current way of doing it
- It avoids dumb mistakes where the user intended to make an exact copy of the other vehicle's orders but forgot to remove any orders currently in the target vehicle's list
- And because it's explicitly intended as a "make this route orders list into an exact copy of that other one" feature, we can more reasonably/confidently do applicable things like also copying the Local/Express option over from the source vehicle

----

And I concur with OP that a feature to "make new copy of vehicle" (in terms of vehicle cars at the very least, and perhaps with orders as well) would be highly desirable. Probably this would be best done as an additional button on the "vehicle details" tab of the vehicle window. Perhaps even as a dropdown menu thing, with one menu item for just cloning the vehicle cars, and another menu item for cloning the vehicle cars *and* the route in one single operation.

One thing we should keep in mind is that we *might* need to remember to do an explicit obsoleteness check on all of the vehicle cars being cloned and deny the clone operation with an error box if any are now obsolete. (This probably depends on the exact methodology we use for cloning; depending on how we go about it, we may or may not hit checks for that which are already in the game. I don't know all the details here.)

**Comment by auge8472** [comment]
As a comparision the way OpenTTD does it:

One can clone a vehicle with it's order. The new vehicle has the same orders as the old one. Changes of the orders of one vehicle does not change the orders of the other one. Two vehicles, two order lists.

One can also clone the vehicle with *sharing* orders (with a shortcut). This might be a special way because of the concrete implementation (I don't know the code but OpenTTD has a limit of order lists for vehicles). That way *a change of the orders of one vehicle changes the orders of all vehicles* that sharing the orders list. That's *very useful* i.e. for a local line with several busses and additionally a shared list is only *one* of the limited order lists *for all* sharing vehicles.

I hope, these thoughts are useful.

**Comment by jgottula** [comment]
> One can also clone the vehicle with sharing orders (with a shortcut). This might be a special way because of the concrete implementation (I don't know the code but OpenTTD has a limit of order lists for vehicles). That way a change of the orders of one vehicle changes the orders of all vehicles that sharing the orders list. That's very useful i.e. for a local line with several busses and additionally a shared list is only one of the limited order lists for all sharing vehicles.

This does sound very handy.

But, as the order list is a persistent property of each vehicle, this definitely sounds like the sort of thing that would need a new/extended save file format before it could be implemented.

**Comment by hemmofa** [comment]
On short term, an option of just cloning a vehicle with or without it's orders would be really welcome!

**Comment by duncanspumpkin** [comment]
We have done enough implementation for this to be added to openloco. I think we should get this feature into the next release.

**Comment by AaronVanGeffen** [comment]
This could be really a interesting new feature. How would you go about implementing it? Would you just be purchasing a new vehicle with the same components, and then setting the orders?

**Comment by duncanspumpkin** [comment]
Yes, will be a case of walking the the cars creating the relevant vehicles copying the build vehicle game commands. Then walking the orders copying them. And finally copying the express/local property. Don't think I'll copy the manual flag.

**Comment by IntelOrca** [comment]
Maybe, have a button on the vehicle window which when you click allows you to place the cloned vehicle as if you were placing a new one.
